### DISCUSSION TITLE
Images with the same Document ID in XMP files are not stacked

### DISCUSSION THREAD

**Discussion by odedia** [photoprism/photoprism#4169] [Q&A] [answered]
#### 1. What is not working as documented?
According to the documentation, photos with the same Document ID will be stacked.
I added a DocumentID to three images as follows:

```
> exiftool -DocumentID="1234-2211" Film0103.nef
    1 image files updated
> exiftool -DocumentID="1234-2211" Film0103.jpg
    1 image files updated
> exiftool -DocumentID="1234-2211" Film0103.edited.jpeg
    1 image files updated
```
The three photos are available for download [here](https://drive.google.com/file/d/1VSjPVfJx0G1IPqgkrlqtS9ekeY-N0N7z/view?usp=sharing).
Confirmed they show up in the metadata using the following command:
```
> exiftool -G -j Film0103.jpg
...
...
  "XMP:DocumentID": "1234-2211",
...
```
I did a complete rescan with cleanup PhotoPrism, but the three images are still not stacked together.
`Film0103.nef` and `Film0103.jpg` are stacked, but `Film0103.edited.jpeg` is shown as a separate photo.

I also tried to use XMP files to achieve this, but the outcome was the same. Here's a sample XMP for reference.

```
<?xpacket begin="﻿" id="W5M0MpCehiHzreSzNTczkc9d"?>
<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="osxphotos">
<rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
<rdf:Description rdf:about="" 
 xmlns:dc="http://purl.org/dc/elements/1.1/" 
 xmlns:photoshop="http://ns.adobe.com/photoshop/1.0/">
        <photoshop:SidecarForExtension>jpeg</photoshop:SidecarForExtension>
        <dc:description>
        <rdf:Alt>
         <rdf:li xml:lang='x-default'></rdf:li>
        </rdf:Alt>
        </dc:description>
        <dc:title>
         <rdf:Alt>
          <rdf:li xml:lang='x-default'/>
         </rdf:Alt>
        </dc:title>
        <dc:subject>
         <rdf:Bag>
          <rdf:li>Blue Sky</rdf:li>
          <rdf:li>Outdoor</rdf:li>
          <rdf:li>People</rdf:li>
          <rdf:li>Rocks</rdf:li>
          <rdf:li>Sky</rdf:li>
         </rdf:Bag>
        </dc:subject>
        <photoshop:DateCreated>1980-01-01T22:05:11.900000+02:00</photoshop:DateCreated>
</rdf:Description>
<rdf:Description rdf:about=""  
 xmlns:Iptc4xmpExt='http://iptc.org/std/Iptc4xmpExt/2008-02-29/'>
</rdf:Description>
<rdf:Description rdf:about="" 
 xmlns:digiKam='http://www.digikam.org/ns/1.0/'>
        <digiKam:TagsList>
        <rdf:Seq>
          <rdf:li>Blue Sky</rdf:li>
          <rdf:li>Outdoor</rdf:li>
          <rdf:li>People</rdf:li>
          <rdf:li>Rocks</rdf:li>
          <rdf:li>Sky</rdf:li>
        </rdf:Seq>
        </digiKam:TagsList>
</rdf:Description>
<rdf:Description rdf:about="" 
 xmlns:xmp='http://ns.adobe.com/xap/1.0/'>
        <xmp:CreateDate>1980-01-01T22:05:11</xmp:CreateDate>
        <xmp:ModifyDate>1980-01-01T22:05:11</xmp:ModifyDate>
        <xmp:DocumentID>1234-2211</xmp:DocumentID>
        <xmp:InstanceID>1234-2211</xmp:InstanceID>
</rdf:Description>
<rdf:Description rdf:about=""
 xmlns:exif='http://ns.adobe.com/exif/1.0/'>
</rdf:Description>
<rdf:Description rdf:about=""
 xmlns:mwg-rs="http://www.metadataworkinggroup.com/schemas/regions/"
 xmlns:stArea="http://ns.adobe.com/xmp/sType/Area#"
 xmlns:stDim="http://ns.adobe.com/xap/1.0/sType/Dimensions#">
</rdf:Description>
<rdf:Description rdf:about=""
 xmlns:MP="http://ns.microsoft.com/photo/1.2/"
 xmlns:MPRI="http://ns.microsoft.com/photo/1.2/t/RegionInfo#"
 xmlns:MPReg="http://ns.microsoft.com/photo/1.2/t/Region#">
</rdf:Description>
</rdf:RDF>
</x:xmpmeta>
<?xpacket end="w"?>
```

**Comment by graciousgrey** [comment, marked as answer]
If you want to group images based on an ID, you can set set the fields "ImageUniqueID" or the "InstanceID" with Exiftool.
We do not yet read these fields from xmp sidecar files.

https://www.photoprism.app/kb/metadata

Since some camera manufacturers or other tools use these fields for serial numbers or similar, we only group files if these fields contain a valid UUID, such as 550e8400-e29b-11d4-a716-446655440000

**Comment by odedia** [comment]
I also set InstanceID with exiftool but that didn’t make a difference. I’ll try with ImageUniqueID and report back. 
also will use a UUID.

**Comment by odedia** [comment]
Setting `InstanceID` did not work, but setting `ImageUniqueID` did work. Thanks!
I suggest updating the docs to reflect exactly which field should be set for Unique ID because right now it seems like DocumentID is a field that you can set.
Thanks again.

**Reply by lastzero** [reply]
`ImageUniqueID` is an Exif field, while `DocumentID` and `InstanceID` are XMP fields, which may also not be supported in external XMP files, but are only embedded:

- In any case, you need to use an ID in a supported format (ideally following the UUID standard that Adobe uses in its tools).
- Setting an `ImageUniqueID` in an unsupported format such as 123 may cause the `DocumentID` to be ignored (even if it is set) and vice versa.
- It's important to be strict here, and we've implemented it this way because users have reported problems otherwise: Some phones simply put their model or serial number in these fields, which caused ALL images to be stacked.

*Note that I am currently preparing our next release and thus only have very limited time to provide in-depth support in our public forums.*

**Reply by odedia** [reply]
To summarise my conclusions from the discussion and my tests for future googlers (would love your confirmation):
- External XMP sidecars are never considered for Unique ID fingerprinting for Stacks
- You can set DocumentID via exiftool such as `exiftool -overwrite_original -DocumentID="51b2fbb7-7d5a-4d41-99c8-e96e1a1a0f25" Film0103.nef`
- You can set ImageUniqueID via exiftool such as `exiftool -overwrite_original -ImageUniqueID="51b2fbb7-7d5a-4d41-99c8-e96e1a1a0f25" Film0103.edited.jpeg`
- - You _cannot_ set InstanceID via exiftool, it appears that PhotoPrism ignores that field when set `exiftool -overwrite_original -InstanceID="51b2fbb7-7d5a-4d41-99c8-e96e1a1a0f25" Film0103.edited.jpeg`

**Reply by lastzero** [reply]
Full support for external XMP files is planned for a later release, as we first need to develop a suitable parser. It would take too long and be too error-prone with the solution we currently have. When XMP is embedded, Exiftool handles the parsing - thus the difference.

**Comment by odedia** [comment]
Is there any way to force an image to be at the top of the stack? Does PhotoPrism look at any field to set order/priority?

**Reply by lastzero** [reply]
No, there is currently no heuristic in place for this - it is only determined by the file naming, i.e. which file is found/indexed first.

**Reply by odedia** [reply]
Thank you. I've seen there's a [feature request](https://github.com/photoprism/photoprism/issues/1028) for this so I'll followup on that one. Stacks is an important feature in my workflow (and for many who edit RAW files), so I would love to see it progress.
Thanks for the great app!
